### PR TITLE
Update setting and displaying of saved interests to use separate interest entity storage.

### PR DESCRIPTION
**Pull request by ellisvivian**

### CODE REVIEW

File: src/main/java/com/google/sps/servlets/BuddyServlet.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    UserService userService = UserServiceFactory.getUserService();
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    String currentUserId = userService.getCurrentUser().getUserId();
+    String otherUserId = request.getParameter(Constants.BUDDY_ADD_PARAM);
+
+    Entity currentUserEntity = getUserEntity(currentUserId);
+    Entity otherUserEntity = getUserEntity(otherUserId);
+
+    List<String> currentBuddiesList = (List<String>) currentUserEntity.getProperty(Constants.USER_BUDDIES_PARAM);
```
Comment (usaeedw) [inline src/main/java/com/google/sps/servlets/BuddyServlet.java]: This line and the line below are wider than 100 characters. Please fix.

Reference: https://google.github.io/styleguide/javaguide.html#s4.4-column-limit

File: src/main/java/com/google/sps/servlets/BuddyServlet.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    List<String> otherBuddiesList = (List<String>) otherUserEntity.getProperty(Constants.USER_BUDDIES_PARAM);
+
+    // If the users are already buddies, remove buddies.
+    if (currentBuddiesList.contains(otherUserId)) {
```
Comment (usaeedw) [inline src/main/java/com/google/sps/servlets/BuddyServlet.java]: I'm not sure I follow this logic. Why remove the users when they are already buddies?

File: src/main/java/com/google/sps/servlets/BuddyServlet.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+      currentBuddiesList.remove(otherUserId);
+      otherBuddiesList.remove(currentUserId);
```
Comment (usaeedw) [inline src/main/java/com/google/sps/servlets/BuddyServlet.java]: Is the otherBuddiesList guaranteed to have `currentUserId`?

File: src/main/java/com/google/sps/servlets/BuddyServlet.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    
+    // If the users are not yet buddies, add buddies.
```
Comment (usaeedw) [inline src/main/java/com/google/sps/servlets/BuddyServlet.java]: Move this comment inside the body of the else. i.e. like: Same for the if-block.

```
else {
  // Some comment.
  ...
}

File: src/main/java/com/google/sps/servlets/BuddyServlet.java (hunk 1 above)
Comment (usaeedw) [inline src/main/java/com/google/sps/servlets/BuddyServlet.java]: Nit: This code below will become more readable if you rename these variables.

An example: `currentUserBuddies` and `otherUserBuddies`. 
 
More generally, avoid the `List` suffix. As long as you are using plurals as your names, no need for adding the `List` suffix.

File: src/main/java/com/google/sps/servlets/UserNameServlet.java [hunk 5]
```diff
@@ -19,62 +19,59 @@
...
+    String id = userService.getCurrentUser().getUserId();
 
-      Entity existingUserEntity = getUserEntity(id);
+    Entity existingUserEntity = getUserEntity(id);
 
-      // If a User entity with the current user's id does not already exist, create it.
-      if (existingUserEntity == null) {
-        Entity newUserEntity = createNewUser(id, name);
-        datastore.put(newUserEntity);
+    // If a User entity with the current user's id does not already exist, create it.
```
Comment (usaeedw) [inline src/main/java/com/google/sps/servlets/UserNameServlet.java]: Move the comments inside the `if` and the `else` blocks.

File: src/main/java/com/google/sps/servlets/InterestServlet.java [hunk 6]
```diff
@@ -16,27 +19,69 @@
...
-    // removes the interest if it has already been saved by the user
-    if (interests.contains(locationName)) {
-      interests.remove(locationName);
+    if (interestEntity == null) {
+      Entity newInterestEntity = createNewInterest(placeId, locationName, userId);
+      datastore.put(newInterestEntity);
+    } else {
+      List<String> interestedUsers = (List<String>) interestEntity.getProperty(Constants.INTEREST_USERS_PARAM);
+      
+      // If the user has already saved this interest, remove them from the list of users.
```
Comment (usaeedw) [inline src/main/java/com/google/sps/servlets/InterestServlet.java]: Same as earlier. Move the comments inside the `if` and `else` blocks.

**Review by usaeedw — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/sps/servlets/BuddyServlet.java (hunk 2 above)
Comment (ellisvivian) [reply src/main/java/com/google/sps/servlets/BuddyServlet.java]: I'm planning on calling this function whenever the user clicks the add or remove buddy button. If the users are already buddies, that means that this function is being called with the intention of removing the buddy relationship. If the users are not yet buddies, that means that this function is being called with the intention of adding the buddy relationship. Is there a different way I should go about things to make it more clear? (I did something similar with the Interest servlet, too)

File: src/main/java/com/google/sps/servlets/BuddyServlet.java (hunk 3 above)
Comment (ellisvivian) [reply src/main/java/com/google/sps/servlets/BuddyServlet.java]: Yes, buddies are always simultaneously added and removed from both lists.

**Review by Omarzintan — COMMENTED** [review]
Just a note here, this might look slightly different after merging my most recent pr that dealt with the proxy url problem. Instead of `result.result.name` it will be `place.name` in the `fetchPlaceInformation()` function in the  `mapScript.js` file 

There's nothing to change here, I just wanted to make you aware of the changes I made to this function

**Review by Omarzintan — APPROVED** [review]

**Review by Omarzintan — APPROVED** [review]

**Review by Omarzintan — APPROVED** [review]

**Review by Omarzintan — APPROVED** [review]

**Review by Omarzintan — APPROVED** [review]

**Review by Omarzintan — APPROVED** [review]

**Review by Omarzintan — APPROVED** [review]

**Review by Omarzintan — DISMISSED** [review]

**Review by Omarzintan — APPROVED** [review]

**Review by Omarzintan — APPROVED** [review]

**Review by Omarzintan — APPROVED** [review]

File: src/main/java/com/google/sps/servlets/BuddyServlet.java (hunk 2 above)
Comment (usaeedw) [reply src/main/java/com/google/sps/servlets/BuddyServlet.java]: That's a good question. Fine for now. 

If you plan to refactor this class, then I would consider making the contract more explicit. Instead using this as a toggle, have two explicit calls for adding and removing a buddy. Even though some of the scaffolding code might be duplicated, in the long run being explicit about state makes the code more readable, easier to reason about, and easier to debug.

File: src/main/java/com/google/sps/servlets/BuddyServlet.java (hunk 3 above)
Comment (usaeedw) [reply src/main/java/com/google/sps/servlets/BuddyServlet.java]: SG. Thanks.

**Review by usaeedw — APPROVED** [review]

**Pull request merged**